### ISSUE TITLE
Best way to include source location annotations in AST or elsewhere?

### ISSUE THREAD

**Issue by davesque** [davesque/rust-vyper#6]
For the past week or two, I've been trying to figure out the most idiomatic and DRY way to include source location information in Rust Vyper's AST.  The motivation here is that, when some kind of compilation error message needs to be generated, the compiler presumably has some piece of the AST it's working with (an instance of an AST-specific struct or enum) and needs to figure out where in the source code this part of the AST was parsed from.  So there needs to be some way to include that info along with or inside of the AST.

So far, I've just been considering ways to include that kind of information *inside* of the AST.

At first, I was just manually declaring a `span: Span` field in all AST-related structs or enum variants as can be seen in [this](https://github.com/davesque/rust-vyper/blob/c11578e8d683e4e65d74c563c8518969cdda8206/parser/src/ast.rs) version of the `ast` module.  Obviously, this is a bit gross and not DRY at all.  You've got to remember to include that field in the struct or enum definition and it also necessitates the existence of methods like [this](https://github.com/davesque/rust-vyper/blob/c11578e8d683e4e65d74c563c8518969cdda8206/parser/src/ast.rs#L163-L172) that resolve the variant of an enum and fetch the `span` field from it.

I initially had a misguided idea of how to fix this, which was to do something like the following:

```rust
struct LocatedNode {
    node: Node,
    span: Span,
}

enum Node {
    Module(Module),
    ModuleStmt(ModuleStmt),
    ConstExpr(ConstExpr),
    ...
}
```

This wouldn't be great because I would need to use this `LocatedNode` type all over the AST and then I would have to resolve the variant of the type at run-time, adding a bunch of computational overhead and eliminating the power of the type system to prevent invalid ASTs from being represented.

Then yesterday I realized I was being a dummy and completely ignoring the basic feature of generic types that Rust provides (not sure why...maybe it's been a while since I've lived in C++ land and places like it).  I ended up just using a struct like this:

```rust
struct Spanned<T> {
    node: T,
    span: Span,
}
```

So now the `ast` module looks like [this](https://github.com/davesque/rust-vyper/blob/d2d43a4745333765a754e4a975634fca11a1459b/parser/src/ast.rs) which is much simpler and more normalized.

I *think* that using this simple parameterized type is the correct way to proceed.  But I still wonder if AST structs and enums (and functions that work on them) can be defined in an even more general way that removes the need to mention the `Spanned` type in the type hierarchy at all.  Maybe AST structs and enums can be defined using generic type parameters somehow or maybe, as I touched on earlier, source location information can be stored in an entirely separate data structure from the AST.

The separate data structure approach might make a lot of sense conceptually since error message generation only happens once in a while and is more of a secondary function of the compiler.  So it doesn't really matter if source location information is conveniently and efficiently available in most cases.

But I don't know, those are just some thoughts.  I think I've basically got a good way to move forward and I think I'm just trying to figure out if there are some other tricks I could use to make things even more maintainable or elegant.

**Comment by davesque** [comment]
@cburgdof @ralexstokes

There's my brain dump as promised.  Please let me know your thoughts on all this or if you need any clarifications.

**Comment by davesque** [comment]
Yesterday, I had the idea of defining the output type of parsers generically with trait bounds.  I thought this might allow me to say statically at a parser's call site what result type I want and, therefore, whether or not the result should include information about spans.  I saved my (incomplete) work in [this](https://github.com/davesque/rust-vyper/commit/d36a42139e96b22248ac7e192e3cbc72a4c7bc8f) commit.

This works, although using it would be impractical.  Not only would I have to define *all* AST types generically with that trait bound, I would also have to specify the desired tree of AST types at a parser's call site.  It might look something like this:

```rust
// Spanned result
let result = file_input::<
    Spanned<Module<'a,
        Spanned<ModuleStmt<'a,
            Spanned<EventField<'a,
                ...,
            >>,
        >>,
    >>,
    SimpleError<_>,
>(input);

// Unspanned result
let result = file_input::<
    Module<'a,
        ModuleStmt<'a,
            EventField<'a,
                ...,
            >,
        >,
    >,
    SimpleError<_>,
>(input);
```

This works because there are implementations for every AST type for the `From<(ASTType, Span)>` trait.  So parsers only say that they'll produce some result which can be converted from and instance of the expected AST type and a `Span` instance that says where it was parsed from.  The `From` implementation for `Spanned` takes the bare type and the span and constructs a `Spanned` instance.  The implementation for un-spanned types just discards the span instance and returns the bare type.

But that's not even a complete picture.  Parsers would have to have a type parameter not just for the immediate output type but for each embedded type inside of the AST.  So the call sites would look like above, but would also include extra concrete types for each embedded type.  In short, it would be a complete nightmare to use.

Anyway, the idea's interesting but unusable.

**Comment by ralexstokes** [comment]
@davesque i would preface w/ "keep it simple" to prioritize moving forward over finding the "best" solution... in this spirit i think the `Spanned<T>` type accomplishes your goals and keeps thing moving.

if you wanted to consider it, i think your idea w/ the "separate data structure makes sense"; AST nodes will still need to be decorated w/ _where_ in the source text they came from but that could simply be indices into a source `str` contained in some kind of `Context` that has e.g. some pseudo-rust:

```rust
struct Context<'s> {
    ast: AST<(probably) 's>,
    source: &'s str,
}
```

then on some error, you would essentially resolve all of this into some nice error data and project to standard out (or wherever you like your error messages) :)

**Comment by davesque** [comment]
Nice, thanks for taking a look!  And I think you're right.  I had envisioned that `Span` instances would only contain byte offsets without saying anything about *what* string is being offset into.  As you say, that would be available in some other scope or data structure.  I think I'll just move forward with that.

**Comment by cburgdorf** [comment]
>I had envisioned that `Span` instances would only contain byte offsets without saying anything about what string is being offset into.

I think that's a good approach to handle it :+1:

**Comment by ralexstokes** [comment]
also worth mentioning to take a look at your favorite rust compiler project(s) and see how they handle this problem

**Comment by davesque** [comment]
@ralexstokes Funny you mention that.  I've actually been diving around quite a bit in the code for the rust compiler itself.  That's where I got the idea for the `Spanned<T>` struct: https://github.com/rust-lang/rust/blob/91642e3ac0120c8e9cdd5f3c85ad03f3bf1b8b69/src/libsyntax_pos/source_map.rs#L43-L46.  Also, strangely, it actually seems that the rust compiler just manually declares a `span` field for every AST related type as can be seen all throughout the `syntax::ast` standard library module: https://github.com/rust-lang/rust/blob/91642e3ac0120c8e9cdd5f3c85ad03f3bf1b8b69/src/libsyntax/ast.rs.  That was the first approach I took, which I had figured was a naive way of doing things.  Anyway, it's been illuminating to look around and see how the rust compiler actually does things.
